<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
<title>IGB 4 Release Notes.</title>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1" />
<meta http-equiv="Content-Language" content="en-US" />
<link rel="stylesheet" type="text/css" media="screen" href="/stylesheet.css" />
<link rel="icon" href="/favicon.ico" />
</head>


<body>
<br>
<h1>IGB 4 release notes.</h1>
<br>
We are pleased to announce the release version <b>4.46</b> of IGB, the Integrated Genome Browser.
<br>
<br>

Highlights:
<ul>

<li><b>Version 4.46</b>  
<ul>

<li>You can output the display to an ".eps" Encapsulated PostScript file.
This allows you to produce high-quality figures for use in publications.
These files can be printed by most laser printers, and can be viewed
with various software, some of which is free.
<br><br><img src="images/print_to_eps.gif">
<br>Figure: Print to an eps file.<br><br>
</li>

<li>The preferences panel for graphs has been removed.  There remain only
a few preferences that can be set by the user and these are in the
"Other Options" panel.
<br><br><img src="images/other_prefs_for_graphs.gif">
<br>Figure: Graph options in the "Other Options" panel.<br><br>
</li>

<li>The splice view now works on a background thread and thus
doesn't slow-down your work in the main view.

<li>The splice view now shows a small "x" mark to indicate where
an exon in some transcript is not being displayed because it occurs
in the intron region of your selected transcript.
<br><br><img src="images/slice_view_with_deletions.gif">
<br>Figure: Slice view showing small "x" marks to indicate missing exons.
Here it was the top transcript that was selected in the main view; it does
not contain exons at the indicated points where the lower transcripts
do contain exons.<br><br>
</li>

<li>In the QuickLoad panel, you are warned before any annotation file
is downloaded over the network.  This helps you avoid downloading
a large file when it is not convenient.

<li>More types of CHP files can be read.....

<li>There is now a "Select All Tiers" menu item in the label map
so that you can apply the same properties to
all selections, such as background color, or collapsed/expanded.
<br><br><img src="images/select_all_tiers.gif">
<br>Figure: Select all tiers to change their properties.<br><br>
</li>

<li>Preferences for QuickLoad and other data access are now available
through the "File->Preferences" menu.

<li>Several features were added to help users who use
genomes other than the ones supported on the NetAffx Quickload
site.
<ul>
<li>You can  provide a name for the genome that is created
when you open a file in a new genome.  For example, you could
load a ".gff" file with the RefSeq data for your genome.
If you later open a graph file that refers to the same genome,
the data will be merged.
<br><br><img src="images/open_dialog.gif">
<br>Figure: Naming a genome during import.<br><br>
</li>
<li>You can specify your own synonyms file.  This is not needed
by most users, but can help when two files refer to the same
genome or chromosome, but use different names for it, such
as "chr1" vs. "Chromosome 1".  Your personal synonyms file
will be merged with any synonyms files that have already been
read, such as the one on the NetAffx QuickLoad server
<a href="http://netaffxdas.affymetrix.com/quickload_data">here</a>, and/or your
own QuickLoad server.  The format is a simple tab-delimited file.
<br><br><img src="images/data_sources_prefs.gif">
<br>Figure: The "data sources" preferences, including the ability
to load a supplementary synonyms file.<br><br>
</li>
</ul>  

<li>There is now a text field that displays the current visible
range of the graph.  You can also quickly go to a region of interest
by typing it into this box.  
Enter coordinates as "start : end", or "start + width" or "center".
<br><br><img src="images/range_box.gif">
<br>Figure: The range box for viewing and going to a specific range.<br><br>
</li>

<li>The autoscroll function now allows the entry of start and end
coordinates.  This is useful when setting-up a background display
for a conference presentation.






<li>Settings and positioning of graphs are now shared across all chromosomes.
For example, if you change the color or range of a graph on chromosome 2, the same
graph on chromosome 3 will also have its color or range changed.

<li>You can now select graphs by selecting either the <em>graph
handle</em> (as in previous versions of IGB) or by selecting the <em>graph
tier label</em>. The graph handle is the small rectangle at the left-edge of
the graph.  The graph tier label is the textual label in the label map
on the left-hand side of the screen.  For non-floating graphs,
the graph tier label is also used to drag the graph tier up or down.  
Floating graphs have no tier label but can
be dragged up and down using the graph handle.

<li>By default, graph labels will now use the short filename rather than
the complete path to the file.  You can change this default through
the preferences panel.

<li>Graph drawing has been improved, including a significant
speed-up in drawing heat-map and bar style graphs.

<li>Graphs for scored-region data, such as expression levels from ".egr" files,
is now drawn in a region-based way which is easier to understand.

<li>Graphs can be joined into a group.  
<ul>
<li>Use the "Join" and "Split" buttons in the "Graph Adjuster" panel.
<li>Grouped graphs can be moved as a unit.  
<li>The group can be "expanded" to show each graph
on a separate line, or "collapsed" to show all graphs printed
on top of each other.
<li>The group can be given a name that differs from the names of individual graphs.
<li>Individual graphs in a group can be manipulated by
by selecting their individual handles.  The group as a whole will
be selected when you select the label in the left-hand label map.
<li>When a graph bookmark is created, the fact that several graphs are
grouped together will be remembered.
</ul>

<li>When an ".egr" file contains multiple scores per region, the
scores will be shown in separate tiers, but those tiers will be
grouped together with a single label in the left-hand label map.
(You can separate these into distinct tiers, if you wish, by using
the "Split" button in the Graph Adjuster tab.)  As with previous
versions of IGB, it is still not yet possible to create a bookmark of
the graphs made from an ".egr" file.

<li>You can change the background color of a graph tier.  
("Floating" graph tiers still have a transparent background.)

<li>The Red/Green color map has been replaced with the more
standard Red/Black/Green color map.

<li>In the graph adjuster and graph threshold setter dialogs, it
is no longer necessary to press Enter after typing a value in a text field.

  
</ul>
  

<br>  
<li><b>Version 4.38</b>

<ul>
<li>The Annotation Browser now allows you to filter the results
to show a chosen subset.  Annotations that map to multiple locations
will list once per location.
<br><br><img src="images/AnnotBrowser2.gif">
<br>Figure: Annotation Browser showing IDs between "a" and "c".<br><br>
</li>

<li>You can sort based on one or more columns in the Annotation Browser.
Click on the column header to set the primary sort column.
Control-click on additional column headers to set secondary sortings.
<br><br><img src="images/AnnotBrowser1.gif">
<br>Figure: Annotation Browser sorted first by Sequence, then by Start.<br><br>
</li>

<li>You can now use the "Get more info" pop-up menu item to get information
on a gene when you have selected an exon or a gene.  It is no longer necessary
to first perform "Select parent" to choose the gene that includes the exon.
Access this pop-up menu by using right-click (control-click on Macintosh)
on the map.
<br><br><img src="images/MoreInfoPopup.gif">
<br>Figure: "Get more info" now works on exons as well as genes.<br><br>
</li>

<li>The Pivot View has been improved.  You can now change the background
and foreground colors, as well as the style and scaling.  Your preferred
choices will be remembered and applied to future sessions.
<br><br><img src="images/PivotView.gif">
<br>Figure: Pivot View now remembers your favorite settings.<br><br>
</li>

<li>It is now possible to "Show" hidden tiers individually by name
in addition to the "Show All" menu item.
<br><br><img src="images/SeqMapViewPopup2.gif">
<br>Figure: You can now show hidden tiers by name.<br><br>
</li>

<li>You can now easily change features of selected tiers with a
"Change" menu item.
<br><br><img src="images/SeqMapViewPopup3.gif">
<br>Figure: The "Change..." menu items.<br><br>
</li>

<li>You can now place the horizontal and vertical zoomers on
whichever sides of the map you prefer.  
Select File-&gt;Preferences-&gt;Other&nbsp;Options.
<br><br><img src="images/ZoomOptions.gif">
<br>Figure: Options for placing the zoom controls.<br><br>
</li>


</ul>



</li><li><b>Version 4.34</b>
<ul>
<li>The "Graph Adjuster" has been redesigned to be easier to use
and it has several new features:
<br><br><img src="images/graph_tab.gif">
<br>Figure: The "Graph Adjuster" tab.<br><br>
<ul>
<li>Options that are not relevant for the current selection will be disabled
</li><li>Graph heights can now be adjusted
</li><li>Additional heat-maps are available, including transparent ones
<br><br><img src="images/transparent_graph.gif">
<br>Figure: A graph with a transparent heat-map<br><br>

</li><li>Graphs may now be re-named (right-click on the tier label)
</li><li>It is possible to combine multiple graphs in one tier (press "Combine" after selecting 2 or more graphs)
</li><li>Graph thresholding is performed in a separate window (press "Thresholding")
<br><br><img src="images/graph_thresholds.gif">
<br>Figure: The graph thresholding dialog.<br><br>
</li></ul>


</li><li>All IGB file types can now be opened in the current genome, or in a new one, with File-&gt;Open
<ul>
<li>By default, "merge" is selected and data will be added to the current genome
</li><li>If "merge" is <em>not</em> selected, a new temporary genome will be created
<br><br><img src="images/file_open.gif"><br>
Figure: The "File-&gt;Open" dialog, with the "merge" checkbox.<br><br>
</li></ul>


</li><li>It is now possible to open Affymetrix files of the "CHP" file type for <em>tiling</em> arrays.  (IGB <em>cannot</em> open "CHP" files from
<em>expression</em> arrays.)
</li><li>You may now access any previously-loaded genome and sequence with the controls on the left-hand side of the "Data Access" tab.<br>
<br><img src="images/current_genome.gif">
<br>Figure: The "Data Access" tab now gives access to all loaded genomes.<br><br>

</li><li>IDs of data from all file types is now indexed so that the IDs
appear in the "Annotation Browser" tab and the search-by-ID function
works for data loaded from all file types </li><li>Multiple items can now be indexed with the same ID
<ul>
<li>Searching for an ID will result in all matching items being selected
</li><li>Searching is no longer case-sensitive
<br><img src="images/multiple_matches_2.gif">
<br>Figure: Searching by ID is not case-sensitive and can find multiple items.<br><br>
</li></ul> 
   
</li><li>The DAS/2 data transfer specification continues to evolve (see
biodas.org). This version of IGB is up-to-date with the current known
DAS/2 servers
</li><li>The interface for data loading from DAS/2 servers has been
improved. Zoom-in to a small region, place a check mark by the items
you'd like to load, then press "Load Features"
<br><img src="images/das2_tab.gif">
<br>Figure: The DAS/2 interface.<br><br>  

</li><li>The "Bookmark Manager" is now opened with "Bookmarks-&gt;Manage Bookmarks"<br>
<br><img src="images/manage_bookmarks.gif">
<br>Figure: To access the "Bookmark Manager", choose "Bookmarks-&gt;Manage Bookmarks".<br><br>

</li></ul>

</li><li><b>Version 4.10</b>:
<ul>
<li>Added support for multiple QuickLoad servers
</li><li>IGB will start-up on the Genome and Sequence you were last looking at
</li><li>The known bugs from version 4.02 have been fixed
</li></ul>

</li><li><b>Version 4.12</b>
<ul>
<li>Making sure to read the das_servers.txt file from QuickLoad servers
</li></ul>

</li><li><b>Version 4.02</b>:
<ul>
<li>User preferences are specified in a graphical interface
</li><li>The <u>igb_prefs.xml</u> file is being deprecated
</li><li>Added support for data-loading from DAS/2 servers
</li><li>Various bug fixes and enhancements
</li></ul>

</li></ul>

<h2>User preference changes</h2>
<h3>The <u>igb_prefs.xml</u> file is being deprecated.</h3>
In versions 3.x, users had to set preferences, such as annotation colors in
a file called <u>igb_prefs.xml</u>.  This file was difficult to edit and did not
allow the changes to be immediately seen.
<br>
<br>
With version 4, the <u>igb_prefs.xml</u> file is almost obsolete.
The majority of users should ignore this file.
(Very advanced users may still find a few uses for it.)
If you are still sometimes using older versions of IGB, you may
choose to keep your <u>igb_prefs.xml</u> file in place.  
Version 4 will simply ignore most of the contents.

<h3>Setting annotation preferences</h3>

When you first load annotation data into IGB, <i>all</i> types of data
will be displayed in the <i>same</i> default color and style.
By default, only "RefSeq" data will display labels.
<br>
<br>
You can change some properties just as before, by right-clicking on
the tier labels and choosing "Hide", "Collapse", "Expand", "Adjust Max Expand",
or "Change Annotation Color".  The main difference now is that these
changes will be persistent.  If you shut down the program and
restart it, these settings will remain.  (Except that hidden tiers
will become visible again when you restart and reload that data.)
<br>
<br>

To access more advanced features, right-click on the tier labels and
choose "Customize".  Alternatively, you can access the same customization
panel by choosing the "File-&gt;Preferences" menu.
<br>
<br>

From the tier preferences panel, you can
<ul>
<li>set foreground and background colors,
</li><li>change the displayed tier names
</li><li>combine (+) and (-) tiers into a single (+/-) tier,
</li><li>collapse or expand tiers,
</li><li>change the maximum tier heights,
</li><li>choose to draw data without connecting lines (appropriate for "repeats" and "contigs"),
</li><li>choose to display labels.
</li></ul>
You may also change the <i>default</i> values of most of these properties.
(A few of the default values are protected from changes.)
For more details, use the "Help" menu on the preferences panel.
Changes take effect as soon as you close this window.
<br>
<br>

Regardless of whether you set properties from the preferences panel
or by right-clicking a tier, the properties will persist between sessions.
Settings for temporarily-created tiers, such as from the union of two
tiers, do not persist between sessions.
<br>
<br>

One other change to note: forward and reverse strands of the same annotation
type now must have matching colors and styles.  Changing one strand automatically
changes the other as well.  (We may later revise this to allow strands to
be hidden independently.)


<h2>Remaining features of <u>igb_prefs.xml</u></h2>

Advanced users may still choose to specify the location of a QuickLoad server
in <u>igb_prefs.xml</u>, or to specify URL's that can be used 
by the "Get More Info" hyperlinks.

<p>
Here is an example of how to set the QuickLoad URL, 
and how to add a hyperlink.
</p>

<code>
&lt;tagval tag="QuickLoadUrl" val="http://netaffxdas.affymetrix.com/quickload_data/"/&gt;
<br>
&lt;annotation_url annot_type_regex="EnsGene"
    url="http://www.ensembl.org/Homo_sapiens/textview?idx=All&amp;q=$$" /&gt;
</code>

<p>
In version 4.02 the URL set this way in the "igb_prefs.xml" file 
will replace the standard NetAffx QuickLoad URL.  (Due to a bug,
any server set with the "Options" button in the QuickLoad panel will
be ignored.)
</p>
<p>
In version 4.10 and higher, you are allowed to have multiple QuickLoad servers.
You will always have access to the NetAffx QuickLoad server,
and you may add an optional "Personal" server using the 
"Options" button in the QuickLoad panel.  To maintain
compatibility with earlier versions, if a QuickLoad URL 
is also set in the "igb_prefs.xml" file, it will become a third,
"Auxiliary" QuickLoad server.
</p>

<h2>DAS/2 servers</h2>

Beginning with version 4.02, there is support for loading data from DAS/2 servers.
<br>
<br>
Earlier versions already had support for loading data from DAS servers.
However, there is a new form of DAS server, DAS version 2.0, that is under development.
Refer to http://biodas.org/ for more information on the DAS/2 specification.
<br>
<br>
Two experimental DAS/2 servers are available now.  Feel free to test them, but
be aware of their experimental nature.  Until the DAS/2 specification becomes
stable, you will generally need the most recent version of IGB to use these
experimental servers.


<h2>DAS/1 servers</h2>
You may continue to access DAS/1 servers as before, using
"File-&gt;Load DAS Features".  
<br>
<br>
If no DAS/1 servers are listed when you try that,
or if you wish to add additional servers,
you may add servers with "File-&gt;Preferences" in the "DAS Servers" panel.
These two servers are a good place to start: 
"http://genome.cse.ucsc.edu/cgi-bin/das/" and 
"http://netaffxdas.affymetrix.com/das/"


<h2>Bug Fixes</h2>

There are various smaller enhancements and bug fixes, including:
<ul>
<li>enhancements and fixes in the "Pivot View"
</li><li>curations have been disabled, but will return in improved form in a future release
</li><li>the Graph Adjuster tab has been simplified somewhat
</li><li>problems with "right-clicking" graphs on Macintosh have been resolved
</li></ul>

<h2>Known Bugs in Version 4.02</h2>

Unfortunately, a few bugs were introduced with release 4.02.
These bugs were fixed in version 4.10.

<p>
<b>QuickLoad server:</b>  The graphical interface for setting the location
of the QuickLoad server <b>does not work</b> in version 4.02.  
If you need to use a server other than the default one, 
you will need to edit your "igb_prefs.xml" file.
</p>
<p>
For example on Linux:<br>
<code>
&lt;tagval tag="QuickLoadUrl" val="file:///home/steve/quickload/" /&gt;
</code>
<br>
or on Windows:
<br>
<code>
&lt;tagval tag="QuickLoadUrl" val="file:///c:/Documents and Settings/name/quickload/" /&gt;
</code>
</p>

<p>
<b>Tier Labels:</b>
If you change the tier label display name through the tier preferences panel,
the name may not take effect until you re-start the program.  Likewise,
if you change from displaying forward and reverse strands separately to displaying them
together, the label may not update to reflect this change, although the data
will be drawn in one or two tiers as you requested.
</p>




<!-- The sourceforge logo is required -->
<a href="http://sourceforge.net"><img src="http://sourceforge.net/sflogo.php?group_id=103871&amp;type=1" alt="SourceForge.net" />

</body>
</html>